### PR TITLE
[LoongArch] Add relax feature and keep relocations

### PR DESCRIPTION
**Pull request by MQ-mengqing**
Add relax feature. To support linker relocation, we should make relocation with a symbol rather than section plus offset, and keep all relocations with non-abs symbol.

### CODE REVIEW

**Comment by MQ-mengqing** [comment]
Add: @SixWeining @wangleiat @xen0n @xry111 @MaskRay .

**Review by SixWeining — COMMENTED** [review]
LGTM. But you can wait for others if you don't mind.

**Comment by MaskRay** [comment]
I am expecting to see something more complete. There is no rush in adding linker relaxation support, and if not careful, can end up needing more efforts to clean up. This is what happened to RISC-V's assembler support, and I finally fixed them after  https://reviews.llvm.org/D153097 https://reviews.llvm.org/D155357 . At the very least, this patch needs more tests, like the tests I improved when improving the two tests.

**Comment by MaskRay** [comment]
Frankly, I am glad that I managed to fix the RISC-V assembler issues before you folks are adding the LoongArch side support... Because otherwise some assembler issue would be more difficult to fix once LoongArch builds dependency on the brittle basis...

**Comment by MQ-mengqing** [comment]
> This is what happened to RISC-V's assembler support, and I finally fixed them after https://reviews.llvm.org/D153097 https://reviews.llvm.org/D155357 . At the very least, this patch needs more tests, like the tests I improved when improving the two tests.

Thank you for the linker relaxation works on RISCV. This allows me to have a reference when I port it  on LoongArch. I noticed these fixes. I think commit [1] relies on both of them. When I supported DwarfLine and DwarfCFA, I found that DwarfCFA did not generate ADD/SUB relocations. Luckily you fixed it for RISCV before.
The current commit is simply to provide relax feature because it will be required for subsequent patches. So it does simple test to see
  a), the symbol is converted to the symbol name instead of section plus offset,
  b), the relocations B16, B21, and B26 are preserved.
Would it be more appropriate to add more tests in later patches?
The smaller connection between patches allows me to submit multiple single-function patches at the same time. If this current commit is merged, I might be able to create PR separately for [1][2][3].

[1] https://github.com/MQ-mengqing/llvm-project/commit/bdcea9b9fe499d9c45e5b8d136b910a6b36aa568
[2] https://github.com/MQ-mengqing/llvm-project/commit/f949e5a3ee6357f35c3a8ea145305f73759f4b74
[3] https://github.com/MQ-mengqing/llvm-project/commit/f6874438ecb8de7c2717518dd2aa7a04f9a1e296

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchELFObjectWriter.cpp [hunk 1]
```diff
@@ -20,19 +20,27 @@ using namespace llvm;
...
+  bool EnableRelax;
 };
 } // end namespace
 
-LoongArchELFObjectWriter::LoongArchELFObjectWriter(uint8_t OSABI, bool Is64Bit)
+LoongArchELFObjectWriter::LoongArchELFObjectWriter(uint8_t OSABI, bool Is64Bit,
+                                                   bool EnableRelax)
     : MCELFObjectTargetWriter(Is64Bit, OSABI, ELF::EM_LOONGARCH,
-                              /*HasRelocationAddend*/ true) {}
+                              /*HasRelocationAddend*/ true),
```
Comment (MaskRay) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchELFObjectWriter.cpp:42]: The canonical comment style is `/*HasRelocationAddend=*/true`

**Review by MaskRay — APPROVED** [review]

**Review by SixWeining — APPROVED** [review]

**Pull request merged**